File: src/query/variable_order/udp_service.py
import socket

class UDPService:
    def __init__(self, local_port: int, receiver_port: int):
        self.receiver_address = ("127.0.0.1", receiver_port)
        self.local_port = local_port  # 本地监听的端口
        self.buffer_size = 1024

        # 创建 UDP 套接字
        self.udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

        # 绑定本地地址到套接字（用于接收消息）
        self.udp_socket.bind(("0.0.0.0", self.local_port))
        print(f"UDP server is listening on port {self.local_port}")

    def receive_message(self):
        data, _ = self.udp_socket.recvfrom(self.buffer_size)
        return data.decode()

    def send_message(self, message):
        self.udp_socket.sendto(message.encode(), self.receiver_address)

File: src/utils/udp_service.cpp
#include "avpjoin/utils/udp_service.hpp"

UDPService::UDPService(int local_port, int receiver_port) {
    // 创建套接字
    if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
        throw std::runtime_error("Socket creation failed.");

    // 配置本地地址（用于接收响应）
    memset(&local_addr, 0, sizeof(local_addr));
    local_addr.sin_family = AF_INET;
    local_addr.sin_addr.s_addr = INADDR_ANY;
    local_addr.sin_port = htons(local_port);

    if (bind(sock, (struct sockaddr*)&local_addr, sizeof(local_addr)) < 0) {
        close(sock);
        throw std::runtime_error("Bind failed.");
    }

    // 配置服务器地址（用于发送消息）
    memset(&receiver_addr, 0, sizeof(receiver_addr));
    receiver_addr.sin_family = AF_INET;
    receiver_addr.sin_port = htons(receiver_port);
    std::string ip = "127.0.0.1";
    if (inet_pton(AF_INET, ip.c_str(), &receiver_addr.sin_addr) <= 0) {
        close(sock);
        throw std::runtime_error("Invalid Service address.");
    }
}

void UDPService::sendMessage(const std::string& message) {
    if (sendto(sock, message.c_str(), message.length(), 0, (struct sockaddr*)&receiver_addr, sizeof(receiver_addr)) <
        0) {
        throw std::runtime_error("Send failed.");
    }
}

std::string UDPService::receiveMessage() {
    char buffer[512];
    memset(buffer, 0, sizeof(buffer));
    int recv_len = recvfrom(sock, buffer, sizeof(buffer) - 1, 0, nullptr, nullptr);
    buffer[recv_len] = '\0';
    return buffer;
}